### DISCUSSION TITLE
ModelSerializer vs Serializer

### DISCUSSION THREAD

**Discussion by AlbertImKr** [weaverse-techtide/weaverse-backend#91] [General]
<h3><strong>비교 요약</strong></h3>

|특성 | ModelSerializer | Serializer
|-- | -- | --
|자동 필드 매핑 | 모델 필드를 자동으로 직렬화 | 수동으로 필드를 정의해야 함
|유효성 검사 | 모델에 정의된 유효성 검사를 자동으로 처리 | 유효성 검사를 직접 정의해야 함
|사용 용도 | 모델 기반의 CRUD API 개발에 적합 | 모델과 관계없이 다양한 데이터 직렬화에 적합
|커스터마이징 | 커스터마이징이 복잡할 수 있음 | 커스터마이징이 쉽고 유연
|개발 속도 | 빠르고 쉽게 직렬화 가능 | 더 많은 코드 작성 필요
|유연성 | 모델과 강하게 연결되어 있어 유연성이 떨어질 수 있음 | 데이터를 자유롭게 직렬화 가능, 모델과 독립적

**Comment by AlbertImKr** [comment]
Serializer를 오직 JSON, XML 등 데이터 형식을 Python으로 변환하거나 Python에서 JSON, XML 등 데이터 형식으로 변환하는 역할을 부여하고자 설계하고 있습니다.

ModelSerializer는 모델 필드를 자동으로 직렬화하는데, 관계 필드가 있는 경우 모델에 접근하여 그 데이터를 가져옵니다. 

```python
class LectureSerializer(serializers.ModelSerializer):
    topics = TopicSerializer(many=True)

    class Meta:
        model = Lecture
        fields = "__all__"


class CourseSerializer(serializers.ModelSerializer):
    lectures = LectureSerializer(
        many=True,
        required=False,
    )

    class Meta:
        model = Course
        fields = "__all__"
```

Course에 Lecture 등 하위 모델이 포함되어 있기 때문에, Course 데이터를 직렬화할 때 Course만 직렬화하고 싶습니다. 하지만 하위 ModelSerializer를 참조하게 되면 모델을 조회하여 예상치 못한 데이터를 보내게 됩니다. 이는 제가 원한 바와 다릅니다.

그래서 저는 Serializer를 사용하는 것이 맞다고 생각합니다.

**Comment by Yoo117** [comment]
ModelSerializer에서도 충분히 원하는 것만 직렬화할 수 있도록 하는 메서드들이 있더라구요. 예를 들면 SerializerMethodField(관계 필드에 대해 ID 목록만 반환), to_representation(관계 필드의 표현을 커스터마이즈) 메서드 오버라이드, depth 옵션 사용(depth = 0  # 관계 필드를 직렬화하지 않음) 등 여러가지 옵션이 있습니다. Serializer를 직접 구현하는 것보다 ModelSerializer를 기반으로 조금씩 커스텀 하는 방식이 코드구현에서 좀 더 간편해 보입니다!

**Comment by AlbertImKr** [comment]
ModelSerializer 사용하기로 했습니다.
